tools/value-generation/gen.py

```diff
@@ -0,0 +1,158 @@
+import yaml
+import copy
+from emitter import CommentedMapping, CommentedDumper
+import sys
+
+# format properties for helm chart
+def format_properties(properties, values, comments, sub_keys, depth):
+  for key, value in properties.items():
+
+    if 'description' not in value:
+      value['description'] = ""
+
+    # Document everything in the YAML but only expose a certain depth in Markdown
+    if depth <= 2:
+      description = '-- ' + value['description']
+    else:
+      description = value['description']
+
+    if 'items' in value:
+      if 'properties' in value['items']:
+        value = value['items']
+
+    # check for sub properties 
+    if 'properties' in value:
+      values[key] = {}
+
+      # place comment key at whatever depth we are in
+      subs = copy.deepcopy(sub_keys)
+      subs.append(key)
+      comments[tuple(subs)] = description
+
+      format_properties(value['properties'], values[key], comments, subs, depth + 1)
+
+    else:
+      comment_key = key 
+      if len(sub_keys):
+        subs = copy.deepcopy(sub_keys)
+        subs.append(key)
+        comment_key = tuple(subs)
+
+      # populate description and default value
+      # comment keys are tuple
+      comments[comment_key] = description
+
+      # Now set the default value: we need this to ensure documentation is generated
+      values[key] = None 
+      if 'default' in value:
+        values[key] = value['default']
+      else:
+        # boolean without default is False
+        if 'type' in value:
+          if value['type'] == 'boolean':
+            values[key] = False
+
+# Set up a lookup table mapping CRD name to Helm chart YAML key
+crd_mapping = {}
+crd_mapping['CouchbaseCluster']='cluster'
+crd_mapping['CouchbaseBucket']='buckets'
+
+# read in crd properties from stdin
+input_crd = sys.stdin.read()
+
+for data in yaml.load_all(input_crd, Loader=yaml.Loader) :
+  crd_name=data['spec']['names']['kind']
+  if crd_name in crd_mapping :
+    crd_value=crd_mapping[crd_name]
+    
+    crd_properties = data['spec']['versions'][0]['schema']['openAPIV3Schema']['properties']['spec']['properties']
+
+    # pass properties into formatter
+    value_map = {}
+    value_map[crd_value] ={}
+    values=value_map[crd_value]
+    comment_map = {}
+    comment_map[crd_value] = '-- Controls the generation of the ' + crd_name + ' CRD'
+    subkeys=[crd_value]
+
+    # Buckets need some special processing
```

Comment:
yea so many 1-offs

Reply:
I've tried to separate this out into its own function now to make it clearer. We probably should look at updating the CRDs with examples for Helm as well so the comments from the CRD are complete for Helm (and others).